### ISSUE TITLE
Browser's built-in zoom isn't honored (ignored) after re-rendering

### ISSUE THREAD

**Issue by Andrew15-5** [Enter-tainer/typst-preview#132]
**Describe the bug**
I want to be able to zoom out a bit in browser to see the whole/more of the page. But as soon as I try to type something after, it will resent/ignore the browser's zoom and reset it as if I never zoomed out.

**To Reproduce**
Steps to reproduce the behavior (Library test):
1. Open preview in browser.
2. Write "Some text" (anything).
3. Zoom out the preview page in the browser.
4. Write a whitespace (anything).
5. Zoom resets and now the page again takes full width of the viewport.

**Expected behavior**
I can see updated preview with zoomed out view without it resetting after each update.

**Package/Software version:**
typst-preview version: `v0.8.1` (https://github.com/Enter-tainer/typst-preview/commit/bf83af63e62515881b192387b6f0e552f9346c02)

**Additional context**

![image](https://github.com/Enter-tainer/typst-preview/assets/37143421/d20612d5-79b9-4d5c-9c81-352414dfaf07)

![image](https://github.com/Enter-tainer/typst-preview/assets/37143421/0c74f94f-83a5-469b-9418-5ee49e5ea444)

![image](https://github.com/Enter-tainer/typst-preview/assets/37143421/ed54adff-39cc-4ce4-b724-5317a839803c)

**Comment by Myriad-Dreamin** [comment]
Could you elaborate a bit? Since we have tested typst-preview on Windows, Ubuntu, Arch, and macOS, and they looks fine to me. So there should be subtle difference between your OS and that operation systems. I remember you are using PopOS, but I don't have one, sadly :(.

**Comment by Andrew15-5** [comment]
Did you test changing browser's tab zoom in between re-renders? Testing just the preview update itself on multiple OSes isn't enough. I can elaborate, but is there something specific that you don't understand? I think I've described/shown everything clear enough.

**Comment by Andrew15-5** [comment]
I've also now tested in clean Firefox profile and ungoogled-chromium (flatpak). The issue persists.

**Comment by Myriad-Dreamin** [comment]
Okay, I believe I understand it, you are using the builtin zoom of browser instead of `ctrl+wheel` provided by typst-preview. It is known to broken since initial release. It should be the time to fix it.

**Comment by Andrew15-5** [comment]
> It should be the time to fix it.

Yay.

> `ctrl+wheel` provided by typst-preview.

First of all, I tried it with a touchpad — it just scrolls the page up and down. Second of all, since the "main" focus is on the VS Code, it means that the majority of the users are GUI fans. But you have to remember about TUI enjoyers who only every use cursor manipulator when every shortcut fails (including Ctrl+-). So yeah, I use Ctrl+- to zoom out.

**Comment by Myriad-Dreamin** [comment]
> touchpad

I did not use touchpad before, but I'm quite curious and tested it just now. It is working on my computer with two fingers pinching in or stretching out. Did you try it?

> So yeah, I use Ctrl+- to zoom out.

Could you share more shortcuts? They must be fit my preference. like the ctrl+- binding is quite useful to many users.

**Comment by Andrew15-5** [comment]
> Could you share more shortcuts?

Are you sure about this? My Neovim/terminal/FOSS OS knowledge of shortcuts is immeasurable. What exactly do you want to know? You can just read the whole https://support.mozilla.org/kb/Keyboard%20shortcuts.

**Comment by Myriad-Dreamin** [comment]
> My Neovim/terminal/FOSS OS knowledge of shortcuts is immeasurable.

Whoops! But we should at least confirm that all of shortcuts from browsers are working, which is just a simpler step.

**Comment by Andrew15-5** [comment]
Well, there is only Ctrl+0, Ctrl+- and Ctrl+= for zoom. It also works in Okular.

**Comment by Andrew15-5** [comment]
> I did not use touchpad before, but I'm quite curious and tested it just now. It is working on my computer with two fingers pinching in or stretching out. Did you try it?

Now, yes. But touchpad a lot of the time isn't supported by default in Linux distros. Pop!_OS has some built-in GNOME extension for touchpad gestures, but I turned it off since it interfered with touchegg touchpad gestures server. In touchegg with empty config nothing will work on pinch, but I added this for firefox:

```xml
  <!-- Zoom in/out in Firefox -->
  <application name="firefox-default">
    <gesture type="PINCH" fingers="2" direction="IN">
      <action type="SEND_KEYS">
        <repeat>true</repeat>
        <modifiers>Control_L</modifiers>
        <keys>KP_Subtract</keys>
        <decreaseKeys>KP_Add</decreaseKeys>
      </action>
    </gesture>
    <gesture type="PINCH" fingers="2" direction="OUT">
      <action type="SEND_KEYS">
        <repeat>true</repeat>
        <modifiers>Control_L</modifiers>
        <keys>KP_Add</keys>
        <decreaseKeys>KP_Subtract</decreaseKeys>
      </action>
    </gesture>
  </application>
```

I know that shindows supports touchpad gestures out of the box.

TL;DR is that there is no positive news, nothing is changed.

**Comment by Myriad-Dreamin** [comment]
Update: typst-preview is a bit joking.
The standalone mode of typst-preview:
+ respect the built-in zoom but ignoring zoom by `ctrl+wheel`.
+ will reset zoom on ctrl+save

Run `Open in browser` command in VSCode:
+ don't like the built-in zoom but only zoom by `ctrl+wheel`.
+ don't reset zoom on ctrl+save

Anyway, both of the issue is highly related to the topic of this issue.

**Comment by Myriad-Dreamin** [comment]
I thought of design a bit, and come to a major problem that: browser cannot remember your scale setting well, since typst-preview uses a random port to serve preview's webpage.

I then have several points to develop the functionality to remember scale setting:
+ Don't remember temporary scale settings, that you have changed during previewing documents.
  Since that one may open preview multiple times and close them in different order, which could cause a weird scale value. For example, close a document with scale `95%` and then close another document with scale `75%`, finally typst-preview may remember the setting: `75%`.
  Please tell me if one have some better idea.

+ Add a `--init-scale` option to cli, and add a corresponding vscode setting `typst-preview.initScale`, whose default value is `100%` or `1`.
  The key is that we should not store the scale setting anywhere, neither on some random configuration file in your personal folders nor in your browser.

+ Add a metadata label `typst-preview-settings-init-scale` to your document, which is then get by `typst query` **only once** during initialization of preview pages.
  The key is that you could store settings and override the configuration at a document level.
  Example:
  ```typst
  // template.typ
  #let my-project-template = [
    #let init-scale = if { get page.width } > a4-paper-width { 0.9 } else { 0.95 }
    metadata({ init-scale }) <typst-preview-settings>
  ]

  // main.typ
  #show: my-project-template
  ```

+ Don't use a random port to serve preview's webpage.

  But this would affect the software structure of typst-preview.

Note: preview with the `100%` scale will adjust the document's width to exact the width of window. preview with the `x%` scale will adjust the document's width to the width of window but left `(1-x)%` unoccupied.
